### PR TITLE
ASoC: Intel: sof_sdw: support different devices on the same sdw link

### PR DESCRIPTION
**Pull request by bardliao**
The existing sdw_sof machine driver construct the dai link name by direction and sdw link id. It means that we will have exactly the same dai link name if two dai links are on the same sdw link with the same direction.
The new Realtek codec has two SoundWire interfacess for jack and DMIC functions and they are treated as different codecs. To create two dai links for jack and DMIC, we need to have different dai link names. This patch suggests to append codec type to the second dai link name. That won't impact the existing topology with this way.

### CODE REVIEW

File: sound/soc/intel/boards/sof_sdw.c [hunk 1]
```diff
@@ -1128,6 +1131,18 @@ static int create_sdw_dailink(struct snd_soc_card *card,
...
 
+		for (index = 0; index < *link_index; index++) {
+			/*
+			 * Append codec type to dai link name if the dai link name is duplicated.
+			 */
+			if (!strcmp(name, dai_links[index].name)) {
+				devm_kfree(dev, name);
+				name = devm_kasprintf(dev, GFP_KERNEL, "%s-%s",
+						      dai_links[index].name,
+						      type_strings[codec_info_list[codec_index].codec_type]);
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c]: commit tile: "ASoC: Intel: sof_sdw: ..."

but this will be confusing, right?
Let's say we have the two DAI on codec0:
1. jack
2. mic

one of the name will be "codec0" the other will be "codec0-mic" or "codec0-jack" depending on the order they are added, so one should be guessing what the "codec0" is?

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -1011,6 +1011,8 @@ static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
 	return 0;
 }
 
+static const char * const type_strings[] = {"jack", "amp", "mic"};
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: probably want to use the SDCA function names, IIRC SimpleJack, SmartAmp, SmartMic (need to double-check)

Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: commit message typo:
The existing sdw_sof machine driver constructs
two SoundWire interfaces

File: sound/soc/intel/boards/sof_sdw.c (hunk 1 above)
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: need to test if (!name)

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c]: @ujfalusi for the current case @bardliao is trying to solve, the dmic and jack are independent SoundWire codecs that happen to be in the same package. So there should be no confusion here.

However when we have to support SDCA, we will need to have multiple functions, so we should probably add a hierarchy such as

````
dailink
    codec[k]
        function[l]
            dai[m]
````

with ALC712, the 'k' value would be 2, 'l' would be 1 and 'm' is 1 as well.

Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: And to @ujfalusi's point there should be a comment that describes the order. Does this match the ACPI tables in sound/soc/intel/common/soc-acpi-...-match.c?

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > And to @ujfalusi's point there should be a comment that describes the order. Does this match the ACPI tables in sound/soc/intel/common/soc-acpi-...-match.c?

@plbossart Yes, the order will match the ACPI tables in sound/soc/intel/common/soc-acpi-...-match.c and the expected order is jack -> amp -> mic. We have a warning message about it. [Unexpected address table ordering. Expected order: jack -> amp -> mic](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw.c#L753)

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: ack

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: fixed

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: ack

**Comment by bardliao** [comment]
Changes: [v2.txt](https://github.com/thesofproject/linux/files/10956826/v2.txt)

File: sound/soc/intel/boards/sof_sdw.c (hunk 1 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > @ujfalusi for the current case @bardliao is trying to solve, the dmic and jack are independent SoundWire codecs that happen to be in the same package. So there should be no confusion here.
> 
> However when we have to support SDCA, we will need to have multiple functions, so we should probably add a hierarchy such as
> 
> ```
> dailink
>     codec[k]
>         function[l]
>             dai[m]
> ```
> 
> with ALC712, the 'k' value would be 2, 'l' would be 1 and 'm' is 1 as well.

@plbossart I have a proposal on https://github.com/thesofproject/linux/pull/4191. The idea is to let dai_name be an array in sof_sdw_codec_info {} and create a dai link for each dai.

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > commit tile: "ASoC: Intel: sof_sdw: ..."
> 
> but this will be confusing, right? Let's say we have the two DAI on codec0:
> 
> 1. jack
> 2. mic
> 
> one of the name will be "codec0" the other will be "codec0-mic" or "codec0-jack" depending on the order they are added, so one should be guessing what the "codec0" is?

@ujfalusi IMHO. we don't rely on the "type" to know the dai link function. There is no "type" in the existing dai link name anyway.  In most cases, the dai link id will be assigned based on its function, that is
```
0: jack out
1: jack in
2: amp
3: amp feedback
4: mic
```

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c]: > @plbossart I have a proposal on #4191. The idea is to let dai_name be an array in sof_sdw_codec_info {} and create a dai link for each dai.

@bardliao I think we could improve a bit with more hierarchy, something like

````
.part_id = 0xdead,
.dai_num = 2,
.dai_descriptor = {
   { 
    	.direction = {true, true},
        .dai_name = "jack-aif1",
	.init = sof_sdw_jack_init,
        .dai_type = SOF_SDW_CODEC_TYPE_JACK,
        .dailink = {0 , 1},
    },
    {
    	.direction = {true, true},
        .dai_name = "amp-aif1",
	.init = sof_sdw_amp_init,
        .dai_type = SOF_SDW_CODEC_TYPE_AMP,
        .dalink = {2, 3},
    },
}
````

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > > @plbossart I have a proposal on #4191. The idea is to let dai_name be an array in sof_sdw_codec_info {} and create a dai link for each dai.
> 
> @bardliao I think we could improve a bit with more hierarchy, something like
> 
> ```
> .part_id = 0xdead,
> .dai_num = 2,
> .dai_descriptor = {
>    { 
>     	.direction = {true, true},
>         .dai_name = "jack-aif1",
> 	.init = sof_sdw_jack_init,
>         .dai_type = SOF_SDW_CODEC_TYPE_JACK,
>         .dailink = {0 , 1},
>     },
>     {
>     	.direction = {true, true},
>         .dai_name = "amp-aif1",
> 	.init = sof_sdw_amp_init,
>         .dai_type = SOF_SDW_CODEC_TYPE_AMP,
>         .dalink = {2, 3},
>     },
> }
> ```

@plbossart .dailink = {0 , 1}, means dai link id = 0 and 1, right? I am wondering do we need to specify the dai link id here.  We arrange dai link id by type. I think it is enough, right?
Another open is that rt712 supports jack and amp function, but customer may use rt712's jack function + another amp like rt1308. How should we handle this configuration? Currently, my idea is to create rt712 amp dai link anyway and topology can ignore the dai link. What do you think?

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c]: Good question @bardliao. My point was that we keep struggling to maintain constant dailink values, depending on configurations we have to change the code to keep the indices as what the topology expects them.

So I was wondering if there's any merit in adding the expected dailink number directly in the descriptor. It's a suggestion only.

I am not sure how we would deal with a case with two amplifier dailinks, one for the 712 amp (not used in practice) and one for the rt1308. There's clearly a conflict with dailink numbers, the topology knows nothing about the codec so which of the two would be selected?

I also wonder if this topology makes any sense. rt712 is mainly for cost-savings with a single-chip solution, OEMs wouldn't add an additional amplifier and not use the RT712 amplifier. What's possible though is that additional amplifiers are used for secondary transducers, but in that case they would be on the same dailink. We probably want to deal with simple configurations first....

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > Good question @bardliao. My point was that we keep struggling to maintain constant dailink values, depending on configurations we have to change the code to keep the indices as what the topology expects them.
> 
> So I was wondering if there's any merit in adding the expected dailink number directly in the descriptor. It's a suggestion only.

Having dailink id in codec_info_list[] is indeed a good idea. My only concern is that someone might design a device with two jack/amp/mic. But it is not likely happened in real world.

> 
> I am not sure how we would deal with a case with two amplifier dailinks, one for the 712 amp (not used in practice) and one for the rt1308. There's clearly a conflict with dailink numbers, the topology knows nothing about the codec so which of the two would be selected?

The amps should be aggregated even they are not the same module, right? 

> 
> I also wonder if this topology makes any sense. rt712 is mainly for cost-savings with a single-chip solution, OEMs wouldn't add an additional amplifier and not use the RT712 amplifier. What's possible though is that additional amplifiers are used for secondary transducers, but in that case they would be on the same dailink. We probably want to deal with simple configurations first....

Agree, let's focus on the known configuration first. I think this PR is good enough to support rt713.

File: sound/soc/intel/boards/sof_sdw.c [hunk 3]
```diff
@@ -1011,6 +1011,8 @@ static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
 	return 0;
 }
 
+static const char * const type_strings[] = {"SimpleJack", "SmartAmp", "SmartMic"};
+
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:1036]: @bardliao I am not following the comment "That won't impact the existing topology with this way."

Are you saying that the existing topology can be used even for this case? I don't see how because the dailinks are different?

Or that we need to generate a new topology where the second dailink name is modified? If yes, can you share the topology change?

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:1036]: > @bardliao I am not following the comment "That won't impact the existing topology with this way."
> 
> Are you saying that the existing topology can be used even for this case? I don't see how because the dailinks are different?
> 
> Or that we need to generate a new topology where the second dailink name is modified? If yes, can you share the topology change?

@plbossart I was trying to say the existing topologies still work with the same HW configurations. Because this commit doesn't change the existing dai link name.
For example, the dai link names are still
```
SDW0-Playback
SDW0-Capture
SDW1-Playback
SDW1-Capture
SDW3-Capture
```
They won't become
```
SDW0-Playback-SimpleJack
SDW0-Capture-SimpleJack
SDW1-Playback-SmartAmp
SDW1-Capture-SmartAmp
SDW3-Capture-SmartMic
```
We only append the type if a previous codec is on the same bus.
For example, If there are two codecs on SDW 0 bus, the dai names will be
```
SDW0-Playback
SDW0-Capture
SDW0-Playback-SmartAmp
SDW0-Capture-SmartAmp
SDW3-Capture
```
The link name will only be changed in such new HW configurations. That's why I said "That won't impact the existing topologies"

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c:1036]: You will still need to modify the topology generation to sometimes use SDW0-Playback-SmartAmp and most times use SDW1-Playback, and in addition the list of dailinks looks weird. How do we know is SDW0-Capture is a jack or a mic?

It's almost like we need to move to a different convention with more explicit names for all links, with some sort of backwards compatibility alias, instead of adding the alias in some cases only.

Need to think more on this one.

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:1036]: > You will still need to modify the topology generation to sometimes use SDW0-Playback-SmartAmp and most times use SDW1-Playback, and in addition the list of dailinks looks weird. How do we know is SDW0-Capture is a jack or a mic?

We only need to add a new topology uses SDW0-Playback-SmartAmp when a product ues jack and amp on the same sdw bus, right? 
For example,
```
"cavs-sdw\;sof-tgl-rt712\;SDW_SPK_STREAM=SDW0-Playback-SmartAmp"
```
The link order is from `struct snd_soc_acpi_link_adr` and we have suggested the jack -> amp -> mic order. Kernel will show a warning if the order doesn't follow the suggestion. So, `SDW0-Capture` before `SDW0-Capture-SmartAmp` should be a jack link.

> 
> It's almost like we need to move to a different convention with more explicit names for all links, with some sort of backwards compatibility alias, instead of adding the alias in some cases only.

That sounds great. But I am not sure do we have backwards compatibility alias. Do you have any idea how to have backwards compatibility alias?

> 
> Need to think more on this one.

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c:1036]: We probably need to ask @ranj063 to remind us how the dailink is bound between machine driver and topology, is this something that we could update with multiple levels and fallbacks?

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:1036]: > We probably need to ask @ranj063 to remind us how the dailink is bound between machine driver and topology, is this something that we could update with multiple levels and fallbacks?

It may work if we can check is there are different codecs on the same sdw bus first, then we can decide what kind of link name format we need.

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:1036]: > > We probably need to ask @ranj063 to remind us how the dailink is bound between machine driver and topology, is this something that we could update with multiple levels and fallbacks?
> 
> It may work if we can check is there are different codecs on the same sdw bus first, then we can decide what kind of link name format we need.

@plbossart Can you review the new version? I implemented my thought. If there are two different codecs on the same sdw link, the driver will append codec type to ALL sdw dai link names.

File: sound/soc/intel/boards/sof_sdw.c [hunk 4]
```diff
@@ -1011,6 +1011,8 @@ static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
 	return 0;
 }
 
+static const char * const type_strings[] = {"SimpleJack", "SmartAmp", "SmartMic"};
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:1035]: commit message: did you mean "This patch suggests to append codec type if there are two or more different types of devices on the same sdw bus"?

We can't change the RVP or Chromebook case with two identical amps on the same link, can we?

File: sound/soc/intel/boards/sof_sdw.c [hunk 5]
```diff
@@ -1021,7 +1023,8 @@ static int create_sdw_dailink(struct snd_soc_card *card,
 			      struct snd_soc_codec_conf *codec_conf,
 			      int codec_count, int *link_id,
 			      int *codec_conf_index,
-			      bool *ignore_pch_dmic)
+			      bool *ignore_pch_dmic,
+			      bool complex_link_name_required)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: "complex" is scary for reviewers :-)

You may want something that describes just the functionality such as 'append_codec_name' or something.

File: sound/soc/intel/boards/sof_sdw.c [hunk 6]
```diff
@@ -1311,8 +1323,20 @@ static int sof_card_dai_links_create(struct device *dev,
 	for (i = 0; i < SDW_MAX_GROUPS; i++)
 		group_generated[i] = false;
 
-	/* generate DAI links by each sdw link */
 	for (; adr_link->num_adr; adr_link++) {
+		/*
+		 * If there is another codec on the same sdw link, we have to append the
+		 * codec type to the dai link name to prevent duplicated dai link name
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: that wouldn't work for the two amplifier on the same link case, would it?

I think you need to test if the part ID and manufacturer ID are different.

**Review by plbossart — COMMENTED** [review]
Sounds good @bardliao, just a couple of details to improve I think.

File: sound/soc/intel/boards/sof_sdw.c (hunk 6 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > that wouldn't work for the two amplifier on the same link case, would it?

We will put the codecs adr in a snd_soc_acpi_adr_device array in this case. I.e. We will put the same codecs in a snd_soc_acpi_adr_device array. 
See
[rt1308_1_dual_adr](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/common/soc-acpi-intel-tgl-match.c#L62)
[mx8373_1_adr](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/common/soc-acpi-intel-tgl-match.c#L131)
So, we won't see another adr_link with the same mask unless there is another codec with different type on the same sdw link.
> 
> I think you need to test if the part ID and manufacturer ID are different.

File: sound/soc/intel/boards/sof_sdw.c (hunk 5 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > "complex" is scary for reviewers :-)
> 
> You may want something that describes just the functionality such as 'append_codec_name' or something.

ack

File: sound/soc/intel/boards/sof_sdw.c (hunk 4 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:1035]: > commit message: did you mean "This patch suggests to append codec type if there are two or more different types of devices on the same sdw bus"?
> 

Yes

> We can't change the RVP or Chromebook case with two identical amps on the same link, can we?

No, we can't.

File: sound/soc/intel/boards/sof_sdw.c (hunk 6 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: @plbossart I tested on a chrome device with dual max98373 on the same sdw link and it works.

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c]: I don't understand your explanations @bardliao. What do you mean by "we won't see another adr_link with the same mask". 

We currently only have a notion of "link_mask" which describe which links are useful/enabled on a given platform. That's used to match with what the platform supports. And then for each adr_link, we will have the list of all devices connected. They could be identical or they could be different. It doesn't matter, they will be on one adr_link.

Maybe you can share what you were planning for the rt712 ACPI tables, that would help me understand what you are suggesting.

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]:
```
static const struct snd_soc_acpi_adr_device rt712_0_single_adr[] = {
	{
		.adr = 0x000030025D071201ull,
		.num_endpoints = 1,
		.endpoints = &single_endpoint,
		.name_prefix = "rt712"
	}
};

static const struct snd_soc_acpi_adr_device rt1712_0_single_adr[] = {
	{
		.adr = 0x000030025D171201ull,
		.num_endpoints = 1,
		.endpoints = &single_endpoint,
		.name_prefix = "rt712-dmic"
	}
};

static const struct snd_soc_acpi_link_adr tgl_712_only[] = {
	{
		.mask = BIT(0),
		.num_adr = ARRAY_SIZE(rt712_0_single_adr),
		.adr_d = rt712_0_single_adr,
	},
	{
		.mask = BIT(0),
		.num_adr = ARRAY_SIZE(rt1712_0_single_adr),
		.adr_d = rt1712_0_single_adr,
	},
	{}
};
```

sof_sdw assumes devices on the same sdw link are all the same. We will need to some more changes on sdw_sof.c if we want to put different devices in a snd_soc_acpi_adr_device array.

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c]: Ah yes, it looks like we have restrictions indeed. I don't think it's quite right to repeat the BIT(0) mask with individual devices. That was not the intent, what I really wanted is one entry per link, and all the devices on that link listed.

That's what I had in mind - different level to add the devices

````c
static const struct snd_soc_acpi_adr_device rt712_0_adr[] = {
	{
		.adr = 0x000030025D071201ull,
		.num_endpoints = 1,
		.endpoints = &single_endpoint,
		.name_prefix = "rt712"
	},
	{
		.adr = 0x000030025D171201ull,
		.num_endpoints = 1,
		.endpoints = &single_endpoint,
		.name_prefix = "rt712-dmic"
	}

};

static const struct snd_soc_acpi_link_adr tgl_712_only[] = {
	{
		.mask = BIT(0),
		.num_adr = ARRAY_SIZE(rt712_0_adr),
		.adr_d = rt712_0_adr,
	},
	{}
};

````

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: @plbossart I updated the code, and we will need some surgery on sof_sdw machine driver to remove the restrictions.

**Comment by keqiaozhang** [comment]
SOFCI TEST

**Comment by bardliao** [comment]
@plbossart I removed the restriction, please have a look. Thanks.

**Review by RanderWang — DISMISSED** [review]
LGTM. Please do some validation for different cases.

File: sound/soc/intel/boards/sof_sdw.c [hunk 7]
```diff
@@ -1033,6 +1033,8 @@ static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
 	return 0;
 }
 
+static const char * const type_strings[] = {"SimpleJack", "SmartAmp", "SmartMic"};
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:1035]: typo in commit message s/interfacess/interfaces/

**Review by plbossart — COMMENTED** [review]
LGTM, only one typo below. We should run a test with firmware 2.2 on all CI devices to see if we catch any issues. 
Also we probably need to see the topology1 changes in parallel with this review so that there's no surprises.
Thanks @bardliao

**Review by RanderWang — APPROVED** [review]

**Review by plbossart — APPROVED** [review]

**Comment by bardliao** [comment]
> We should run a test with firmware 2.2 on all CI devices to see if we catch any issues. 

I triggered a test with firmware 2.2 (#23328), and got a positive result.
